docs/DeveloperGuide.md

```diff
@@ -1,34 +1,972 @@
+
+##### 3.4.2.1 `ModuleList` Class
+
+The `ModuleList` class acts as an abstraction for the ArrayList that is used to store module objects created from any of the 
+classes under the `module` package. It provides a `ModuleList` structure with features built on top of ArrayList to enhance and 
+customize its usage for the application. Some data-related methods such as `addModuleRequiredBy` and `removeTakenPreRequisiteModule` 
+are added to aid data processing when adding a module to the `ModuleList`. The `ModuleList` class is also used by the `Storage` 
+component to generate JSON data of all modules and store it to disk.
+
+Within the class `ModuleList`, different methods are defined to perform different operations on the list of `Modules`. These
+operations are:
+1. Add a module to the list
+2. Delete a module from the list
+3. Mark a module as done
+
+##### Adding a module
+
+For adding modules to the list, the methods defined are `add`, `addModuleRequiredBy` and `removeTakenByPrerequisites`.
+To add a new module, say `newModule`, `add` is called to perform 3 steps:
+1. First, `addModuleRequiredBy`is called to populate the list of modules that require
+`newModule` as a prerequisite. 
+2. Then, `removeTakenByPrerequisites` checks the list of prerequisites of `newModule` and 
+removes the ones that have been marked as taken.
+3. Finally, `add` calls `ArrayList.add()` to add `newModule` to the list.
+
+##### Deleting a module
+
+For deleting a module to the list, the methods defined are `delete` and `removeFromPreRequisiteModuleRequiredBy`. To
+delete an existing module from the list, say `existingModule`, `delete` is called to perform 2 steps:
+1. Firstly, `delete` calls `ArrayList.remove()` to delete `existingModule` from the list.
+2. Then, `removeFromPreRequisiteModuleRequiredBy` is called to remove `existingModule` from its pre-requisite modules' 
+   requiredBy list.
+   
+##### Mark module as taken
+
+For marking a module as taken, the methods defined are `markAsTaken` and `removeFromModuleUntakenPrerequisites`. To mark
+an existing module as taken, say `existingModule`, `markAsTaken` is called to perform 2 steps:
+1. First, `markAsTaken` calls `Module.setStatus` and sets the status of `existingModule` to "taken".
+2. Then, `removeFromModuleUntakenPrerequisites` is called to remove `existingModule` from the prerequisitesUntaken table
+from the list of modules that require `existingModule` as a prerequisite.
+
+Given below is the class diagram of `ModuleList` showing its 3 main functions.
+
+![archi](./images/ModuleListClassDiagram.png)
+
+<sup>***Figure 3.4.2.1** UML class diagram for ModuleList class*</sup>
+
+Apart from these 3 operations, the `ModuleList` class also defines getter and setter methods to retrieve values such
+as the entire list or an individual module from the list according to different parameters such as module code or index. 
+
+### 3.5 Storage Component
+
+Class Diagram:
+
+![archi](./images/storageClassDiagram.jpg)
```

Comment:
Small issue but do you think it's better to use the same uml diagram style throughout the document?